climpred/classes.py

```diff
@@ -0,0 +1,197 @@
+import xarray as xr
+from .prediction import compute_reference, compute_persistence
+# TODO: add horizon functionality
+# TODO: add perfect model functionality
+# TODO: add relative entropy functionality
+# TODO: add various `get` and `set` decorators
+# TODO: refactor other modules, e.g. move all metrics to `metrics.py`
+
+
+# --------------
+# VARIOUS CHECKS
+# --------------
+def _check_prediction_ensemble_dimensions(xobj):
+    """
+    Checks that at the minimum, the dple object has dimensions initialization
+    and time (a time series with lead times).
+    """
+    cond = all(dims in xobj.dims for dims in ['initialization', 'time'])
+    if not cond:
+        # create custom error here.
+        raise ValueError("""Your decadal prediction object must contain the
+            dimensions `time` and `initialization` at the minimum.""")
+
+
+def _check_reference_dimensions(init, ref):
+    """Checks that the reference matches all initialized dimensions except
+    for 'time'"""
+    init_dims = list(init.dims)
+    if 'time' in init_dims:
+        init_dims.remove('time')
+    if not (set(ref.dims) == set(init_dims)):
+        raise ValueError("""Reference dimensions must match initialized
+            prediction ensemble dimensions (excluding `time`.)""")
+
+
+# TODO: Don't force same variables for every reference. Allow for some
+# references to only have a subset of the initialized prediction variables.
+# This means skill, etc. will be computed for whatever variables
+# are available.
+def _check_reference_vars_match_initialized(init, ref):
+    """
+    Checks that a new reference (or control) dataset has identical variables
+    to the initialized dataset. This ensures that they can be compared
+    pairwise.
+
+    ref: new addition
+    init: dp.initialized
+    """
+    init_list = [var for var in init.data_vars]
+    ref_list = [var for var in ref.data_vars]
+    if set(init_list) != set(ref_list):
+        raise ValueError("""Please provide a dataset with matching variables
+        (and variable names) to the initialied prediction ensemble.""")
+
+
+def _check_xarray(xobj):
+    if not isinstance(xobj, (xr.Dataset, xr.DataArray)):
+        raise ValueError("""You must input an xarray Dataset or DataArray.""")
+
+
+# ----------
+# Aesthetics
+# ----------
+def _display_metadata(self):
```

Comment:
general comment/thought: with xarray the metadata vanish at some point. the `esmlab` people try to update the metadata after their functions. `cdo` also adds all operations. I check those with `ncdump -h file`

Reply:
We can talk about this on our skype call. The `__repr__` function is a special python function that basically is called when you type `print(dp)` or `dp` in Jupyter. So we can make this whatever we want.

I am currently just printing out the variables that the user has in each dataset to make it cleaner while developing (so that we don't have all the coordinates and everything tracking along). But we can talk about what would be ideal. I think attributes/metadata should come along with it as well.